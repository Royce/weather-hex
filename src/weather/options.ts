import _ from "lodash";

import { uniquePairs } from "./uniquePairs";

export type Temperature = "hot" | "warm" | "cool" | "freezing";
const temperatureScale: Temperature[] = ["hot", "warm", "cool", "freezing"];

export type Wind = "calm" | "breeze" | "gale";
const windScale: Wind[] = ["calm", "breeze", "gale"];

export type Sky = "overcast" | "light clouds" | "clear";
const skyScale: Sky[] = ["overcast", "light clouds", "clear"];

export type Water =
  | "light rain"
  | "heavy rain"
  | "hail"
  | "light snowfall"
  | "heavy snowfall"
  | "lightning"
  | "frost"
  | "fog"
  | "humid"
  | "dry";
const waterList: Water[] = [
  "light rain",
  "heavy rain",
  "hail",
  "light snowfall",
  "heavy snowfall",
  "lightning",
  "frost",
  "fog",
  "humid",
  "dry",
];

export type WeatherValue = Temperature | Wind | Sky | Water;

const requires: { [K in Water]?: WeatherValue[] } = {
  hail: ["cool", "freezing"],
  "light snowfall": ["freezing"],
  "heavy snowfall": ["freezing"],
  humid: ["hot", "warm"],
  frost: ["freezing"],
  fog: ["cool"],
  lightning: ["overcast"],
};

const excludes: { [K in Water]?: WeatherValue[] } = {
  "light rain": ["clear", "freezing", "heavy rain"],
  "heavy rain": ["hot", "clear", "freezing"],
  hail: ["clear"],
  "light snowfall": ["clear"],
  "heavy snowfall": ["clear", "light clouds"],
  lightning: ["calm", "freezing", "hot"],
  fog: ["breeze", "gale", "heavy rain"],
  humid: ["gale"],
};

const exclusive: Water[] = ["dry", "light snowfall", "heavy snowfall"];

export type Weather = {
  temperature: Temperature;
  wind: Wind;
  sky: Sky;
  water: Water[];
};

function ok({ temperature, wind, sky, water: waterList }: Weather): boolean {
  const matches = (test: WeatherValue) =>
    test === temperature ||
    test === wind ||
    test === sky ||
    _.includes(waterList, test);

  return _.every(waterList, (water) => {
    return (
      (excludes[water] === undefined ||
        _.every(excludes[water], (t) => !matches(t))) &&
      (requires[water] === undefined || _.some(requires[water], matches))
    );
  });
}

export function all() {
  var list: Weather[] = [];

  for (const temperature of temperatureScale) {
    for (const wind of windScale) {
      for (const sky of skyScale) {
        for (const water of exclusive) {
          const weather = { temperature, wind, sky, water: [water] };
          if (ok(weather)) list.push(weather);
        }
        for (const water of _.without(waterList, ...exclusive)) {
          const weather = { temperature, wind, sky, water: [water] };
          if (ok(weather)) list.push(weather);
        }
        for (const water of uniquePairs(_.without(waterList, ...exclusive))) {
          const weather = { temperature, wind, sky, water };
          if (ok(weather)) list.push(weather);
        }
      }
    }
  }

  return list;
}
