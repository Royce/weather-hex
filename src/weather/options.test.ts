import { all } from "./options";

test("generate all useful weather permutations", () => {
  const options = all();

  expect(
    JSON.stringify(
      options
        .map(
          ({ temperature, wind, sky, water }) =>
            `${water.join(", ")}: ${temperature}, ${wind}, ${sky}`
        )
        .sort(),
      null,
      2
    )
  ).toMatchSnapshot();
});
