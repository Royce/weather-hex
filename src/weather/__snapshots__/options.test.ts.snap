// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate all useful weather permutations 1`] = `
"[
  \\"dry: cool, breeze, clear\\",
  \\"dry: cool, breeze, light clouds\\",
  \\"dry: cool, breeze, overcast\\",
  \\"dry: cool, calm, clear\\",
  \\"dry: cool, calm, light clouds\\",
  \\"dry: cool, calm, overcast\\",
  \\"dry: cool, gale, clear\\",
  \\"dry: cool, gale, light clouds\\",
  \\"dry: cool, gale, overcast\\",
  \\"dry: freezing, breeze, clear\\",
  \\"dry: freezing, breeze, light clouds\\",
  \\"dry: freezing, breeze, overcast\\",
  \\"dry: freezing, calm, clear\\",
  \\"dry: freezing, calm, light clouds\\",
  \\"dry: freezing, calm, overcast\\",
  \\"dry: freezing, gale, clear\\",
  \\"dry: freezing, gale, light clouds\\",
  \\"dry: freezing, gale, overcast\\",
  \\"dry: hot, breeze, clear\\",
  \\"dry: hot, breeze, light clouds\\",
  \\"dry: hot, breeze, overcast\\",
  \\"dry: hot, calm, clear\\",
  \\"dry: hot, calm, light clouds\\",
  \\"dry: hot, calm, overcast\\",
  \\"dry: hot, gale, clear\\",
  \\"dry: hot, gale, light clouds\\",
  \\"dry: hot, gale, overcast\\",
  \\"dry: warm, breeze, clear\\",
  \\"dry: warm, breeze, light clouds\\",
  \\"dry: warm, breeze, overcast\\",
  \\"dry: warm, calm, clear\\",
  \\"dry: warm, calm, light clouds\\",
  \\"dry: warm, calm, overcast\\",
  \\"dry: warm, gale, clear\\",
  \\"dry: warm, gale, light clouds\\",
  \\"dry: warm, gale, overcast\\",
  \\"fog: cool, calm, clear\\",
  \\"fog: cool, calm, light clouds\\",
  \\"fog: cool, calm, overcast\\",
  \\"frost: freezing, breeze, clear\\",
  \\"frost: freezing, breeze, light clouds\\",
  \\"frost: freezing, breeze, overcast\\",
  \\"frost: freezing, calm, clear\\",
  \\"frost: freezing, calm, light clouds\\",
  \\"frost: freezing, calm, overcast\\",
  \\"frost: freezing, gale, clear\\",
  \\"frost: freezing, gale, light clouds\\",
  \\"frost: freezing, gale, overcast\\",
  \\"hail, fog: cool, calm, light clouds\\",
  \\"hail, fog: cool, calm, overcast\\",
  \\"hail, frost: freezing, breeze, light clouds\\",
  \\"hail, frost: freezing, breeze, overcast\\",
  \\"hail, frost: freezing, calm, light clouds\\",
  \\"hail, frost: freezing, calm, overcast\\",
  \\"hail, frost: freezing, gale, light clouds\\",
  \\"hail, frost: freezing, gale, overcast\\",
  \\"hail, lightning: cool, breeze, overcast\\",
  \\"hail, lightning: cool, gale, overcast\\",
  \\"hail: cool, breeze, light clouds\\",
  \\"hail: cool, breeze, overcast\\",
  \\"hail: cool, calm, light clouds\\",
  \\"hail: cool, calm, overcast\\",
  \\"hail: cool, gale, light clouds\\",
  \\"hail: cool, gale, overcast\\",
  \\"hail: freezing, breeze, light clouds\\",
  \\"hail: freezing, breeze, overcast\\",
  \\"hail: freezing, calm, light clouds\\",
  \\"hail: freezing, calm, overcast\\",
  \\"hail: freezing, gale, light clouds\\",
  \\"hail: freezing, gale, overcast\\",
  \\"heavy rain, hail: cool, breeze, light clouds\\",
  \\"heavy rain, hail: cool, breeze, overcast\\",
  \\"heavy rain, hail: cool, calm, light clouds\\",
  \\"heavy rain, hail: cool, calm, overcast\\",
  \\"heavy rain, hail: cool, gale, light clouds\\",
  \\"heavy rain, hail: cool, gale, overcast\\",
  \\"heavy rain, humid: warm, breeze, light clouds\\",
  \\"heavy rain, humid: warm, breeze, overcast\\",
  \\"heavy rain, humid: warm, calm, light clouds\\",
  \\"heavy rain, humid: warm, calm, overcast\\",
  \\"heavy rain, lightning: cool, breeze, overcast\\",
  \\"heavy rain, lightning: cool, gale, overcast\\",
  \\"heavy rain, lightning: warm, breeze, overcast\\",
  \\"heavy rain, lightning: warm, gale, overcast\\",
  \\"heavy rain: cool, breeze, light clouds\\",
  \\"heavy rain: cool, breeze, overcast\\",
  \\"heavy rain: cool, calm, light clouds\\",
  \\"heavy rain: cool, calm, overcast\\",
  \\"heavy rain: cool, gale, light clouds\\",
  \\"heavy rain: cool, gale, overcast\\",
  \\"heavy rain: warm, breeze, light clouds\\",
  \\"heavy rain: warm, breeze, overcast\\",
  \\"heavy rain: warm, calm, light clouds\\",
  \\"heavy rain: warm, calm, overcast\\",
  \\"heavy rain: warm, gale, light clouds\\",
  \\"heavy rain: warm, gale, overcast\\",
  \\"heavy snowfall: freezing, breeze, overcast\\",
  \\"heavy snowfall: freezing, calm, overcast\\",
  \\"heavy snowfall: freezing, gale, overcast\\",
  \\"humid: hot, breeze, clear\\",
  \\"humid: hot, breeze, light clouds\\",
  \\"humid: hot, breeze, overcast\\",
  \\"humid: hot, calm, clear\\",
  \\"humid: hot, calm, light clouds\\",
  \\"humid: hot, calm, overcast\\",
  \\"humid: warm, breeze, clear\\",
  \\"humid: warm, breeze, light clouds\\",
  \\"humid: warm, breeze, overcast\\",
  \\"humid: warm, calm, clear\\",
  \\"humid: warm, calm, light clouds\\",
  \\"humid: warm, calm, overcast\\",
  \\"light rain, fog: cool, calm, light clouds\\",
  \\"light rain, fog: cool, calm, overcast\\",
  \\"light rain, hail: cool, breeze, light clouds\\",
  \\"light rain, hail: cool, breeze, overcast\\",
  \\"light rain, hail: cool, calm, light clouds\\",
  \\"light rain, hail: cool, calm, overcast\\",
  \\"light rain, hail: cool, gale, light clouds\\",
  \\"light rain, hail: cool, gale, overcast\\",
  \\"light rain, humid: hot, breeze, light clouds\\",
  \\"light rain, humid: hot, breeze, overcast\\",
  \\"light rain, humid: hot, calm, light clouds\\",
  \\"light rain, humid: hot, calm, overcast\\",
  \\"light rain, humid: warm, breeze, light clouds\\",
  \\"light rain, humid: warm, breeze, overcast\\",
  \\"light rain, humid: warm, calm, light clouds\\",
  \\"light rain, humid: warm, calm, overcast\\",
  \\"light rain, lightning: cool, breeze, overcast\\",
  \\"light rain, lightning: cool, gale, overcast\\",
  \\"light rain, lightning: warm, breeze, overcast\\",
  \\"light rain, lightning: warm, gale, overcast\\",
  \\"light rain: cool, breeze, light clouds\\",
  \\"light rain: cool, breeze, overcast\\",
  \\"light rain: cool, calm, light clouds\\",
  \\"light rain: cool, calm, overcast\\",
  \\"light rain: cool, gale, light clouds\\",
  \\"light rain: cool, gale, overcast\\",
  \\"light rain: hot, breeze, light clouds\\",
  \\"light rain: hot, breeze, overcast\\",
  \\"light rain: hot, calm, light clouds\\",
  \\"light rain: hot, calm, overcast\\",
  \\"light rain: hot, gale, light clouds\\",
  \\"light rain: hot, gale, overcast\\",
  \\"light rain: warm, breeze, light clouds\\",
  \\"light rain: warm, breeze, overcast\\",
  \\"light rain: warm, calm, light clouds\\",
  \\"light rain: warm, calm, overcast\\",
  \\"light rain: warm, gale, light clouds\\",
  \\"light rain: warm, gale, overcast\\",
  \\"light snowfall: freezing, breeze, light clouds\\",
  \\"light snowfall: freezing, breeze, overcast\\",
  \\"light snowfall: freezing, calm, light clouds\\",
  \\"light snowfall: freezing, calm, overcast\\",
  \\"light snowfall: freezing, gale, light clouds\\",
  \\"light snowfall: freezing, gale, overcast\\",
  \\"lightning, humid: warm, breeze, overcast\\",
  \\"lightning: cool, breeze, overcast\\",
  \\"lightning: cool, gale, overcast\\",
  \\"lightning: warm, breeze, overcast\\",
  \\"lightning: warm, gale, overcast\\"
]"
`;
